--- mariadb-10.4.10/sql/item_subselect.cc	2019-11-07 14:52:22.000000000 +0900
+++ mariadb-10.4.10.patch/sql/item_subselect.cc	2019-11-19 12:02:21.860540738 +0900
@@ -123,7 +123,7 @@
                     NO_MATTER :
                     outer_select->parsing_place);
     if (unit->is_unit_op() &&
-        (unit->first_select()->next_select() or unit->fake_select_lex))
+        (unit->first_select()->next_select() || unit->fake_select_lex))
       engine= new subselect_union_engine(unit, result, this);
     else
       engine= new subselect_single_select_engine(select_lex, result, this);
