diff -ru mariadb-10.0.21-with-mroonga-5.06.orig/storage/mroonga/vendor/groonga/lib/grn.h mariadb-10.0.21-with-mroonga-5.06/storage/mroonga/vendor/groonga/lib/grn.h
--- mariadb-10.0.21-with-mroonga-5.06.orig/storage/mroonga/vendor/groonga/lib/grn.h	2015-08-13 18:15:38.000000000 +0900
+++ mariadb-10.0.21-with-mroonga-5.06/storage/mroonga/vendor/groonga/lib/grn.h	2015-08-31 17:13:59.251028024 +0900
@@ -47,6 +47,10 @@
 #  undef NTDDI_VERSION
 # endif /* NTDDI_VERSION */
 # define NTDDI_VERSION GRN_MINIMUM_WINDOWS_VERSION
+
+# ifdef WIN32_LEAN_AND_MEAN
+#  undef WIN32_LEAN_AND_MEAN
+# endif /* WIN32_LEAN_AND_MEAN */
 #endif /* WIN32 */
 
 #ifdef __cplusplus
