--- mariadb-10.5.5.orig/sql/sql_type.h	2020-09-09 13:20:40.113007137 +0900
+++ mariadb-10.5.5/sql/sql_type.h	2020-09-09 13:23:17.385795235 +0900
@@ -7298,8 +7298,8 @@ extern MYSQL_PLUGIN_IMPORT Named_type_ha
 
 extern Named_type_handler<Type_handler_bit>         type_handler_bit;
 
-extern Named_type_handler<Type_handler_enum>        type_handler_enum;
-extern Named_type_handler<Type_handler_set>         type_handler_set;
+extern MYSQL_PLUGIN_IMPORT Named_type_handler<Type_handler_enum>        type_handler_enum;
+extern MYSQL_PLUGIN_IMPORT Named_type_handler<Type_handler_set>         type_handler_set;
 
 extern Named_type_handler<Type_handler_string>      type_handler_string;
 extern Named_type_handler<Type_handler_var_string>  type_handler_var_string;
